rust/lance/src/io/exec/optimizer.rs

```diff
@@ -0,0 +1,57 @@
+// Copyright 2023 Lance Developers.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! Lance Physical Optimizer Rules
+
+use std::sync::Arc;
+
+use datafusion::{
+    common::tree_node::{Transformed, TreeNode},
+    config::ConfigOptions,
+    error::Result as DFResult,
+    physical_optimizer::PhysicalOptimizerRule,
+    physical_plan::ExecutionPlan,
+};
+
+use super::TakeExec;
+
+/// Rule that eliminates [TakeExec] nodes that are immediately followed by another [TakeExec].
```

Comment:
Wouldn't this be an invalid input plan?  Doesn't `TakeExec` need row ids as input?  It doesn't output row ids does it?

Reply:
It almost always outputs row ids. For example:

https://github.com/lancedb/lance/blob/ee1fb9aa8e4bd4989678588889ecc916211a6cd9/rust/lance/src/dataset/scanner.rs#L3350-L3353